File: HW/3_Quera/1_poshte_faza'ee.py
class Stack:
    def __init__(self):
        self.list = []
        self.minimum = 0

    def push(self, element):
        if self.is_empty:
            self.minimum = element
            self.list.append(element)
        else:
            if self.minimum <= element:
                self.list.append(element)
            else:
                self.minimum = element
                self.list.append(element)

    def pop(self):
        return self.list.pop()

    def spell(self):
        return self.minimum

    def is_empty(self):
        if self.list.__len__() == 0:
            return True
        else:
            return False


stack = Stack()
q = int(input())
commands = [''] * q
for i in range(q):
    commands[i] = input()

for j in range(q):
    if commands[j].__contains__("push"):
        push_array = commands[j].split()
        stack.push(push_array[1])
        print(push_array[1])
    if commands[j].__contains__("pop"):
        # print("pop " + stack.pop())
        stack.pop()
    if commands[j].__contains__("spell"):
        print("spell " + stack.spell())

File: HW/3_Quera/Q1/1_poshte_faza'ee.java
import java.util.ArrayList;
import java.util.Scanner;
import java.util.Stack;

public class Main {

        public static void main(String[] args) {
            Stack<Pair> stack = new Stack<>();
            Scanner in = new Scanner(System.in);

            int operationNumber = Integer.parseInt(in.nextLine());
            ArrayList<String> operations = new ArrayList<>();
            for (int i = 0; i < operationNumber; i++) {
                operations.add(in.nextLine());
            }

            for(int i = 0 ; i < operationNumber; i++) {

                if (operations.get(i).equals("pop")) {
                    if(!stack.isEmpty())
                        stack.pop();
                }
                else if (operations.get(i).contains("push")) {
                    int x = Integer.parseInt(operations.get(i).substring(5));
                    if(stack.isEmpty())
                        stack.push(new Pair(x, x));
                    else
                        stack.push(new Pair(x, Math.min(x, stack.peek().minimum)));

                }
                else if (operations.get(i).equals("spell")) {
                    if(!stack.isEmpty())
                        System.out.println(stack.peek().minimum);
                }

            }
        }

        private static class Pair {
            int value, minimum;

            public Pair(int val, int min) {
                value = val;
                minimum = min;
            }
        }
    }

